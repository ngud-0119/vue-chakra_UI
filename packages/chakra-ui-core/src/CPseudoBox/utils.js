import map from 'lodash-es/map'
import { forwardProps, filterBaseStyles, filterPseudo as getPseudoStyles, tx } from '../utils'

/**
 * PseudoBox pseudo selectors
 */
export const selectors = {
  _hover: '&:hover',
  _active: '&:active, &[data-active=true]',
  _focus: '&:focus',
  _visited: '&:visited',
  _even: '&:nth-of-type(even)',
  _odd: '&:nth-of-type(odd)',
  _disabled: '&:disabled, &:disabled:focus, &:disabled:hover, &[aria-disabled=true], &[aria-disabled=true]:focus, &[aria-disabled=true]:hover',
  _checked: '&[aria-checked=true]',
  _mixed: '&[aria-checked=mixed]',
  _selected: '&[aria-selected=true]',
  _invalid: '&[aria-invalid=true]',
  _pressed: '&[aria-pressed=true]',
  _readOnly: '&[aria-readonly=true], &[readonly]',
  _first: '&:first-of-type',
  _last: '&:last-of-type',
  _expanded: '&[aria-expanded=true]',
  _grabbed: '&[aria-grabbed=true]',
  _notFirst: '&:not(:first-of-type)',
  _notLast: '&:not(:last-of-type)',
  _groupHover: '[role=group]:hover &',
  _before: '&:before',
  _after: '&:after',
  _focusWithin: '&:focus-within',
  _placeholder: '&::placeholder'
}

/**
 * Filter undefined props and parse pseudo style props to generate css styles object
 * @param {Object} props - Props
 * @returns {Object} Parsed pseudo styles object
 */
export function parsePseudoStyles (props) {
  const pseudoStyles = {}
  const clean = forwardProps(props)
  const pseudoProps = getPseudoStyles(clean)
  const baseProps = filterBaseStyles(clean)
  const result = map(pseudoProps, (value, prop) => ({ prop, value }))
  result.forEach((pair) => {
    if (selectors[pair.prop]) {
      pseudoStyles[selectors[pair.prop]] = tx(pair.value)
    }
  })
  const mergedPseudoBoxStyles = { pseudoStyles, baseProps }
  return mergedPseudoBoxStyles
}
