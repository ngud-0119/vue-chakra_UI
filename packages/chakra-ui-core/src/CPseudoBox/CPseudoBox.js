import { css } from 'emotion'
import __css from '@styled-system/css'
import { background, border, color, borderRadius, flexbox, grid, layout, position, shadow, space, typography, compose } from 'styled-system'
import CBox from '../CBox'
import styleProps, { propsConfig } from '../config/props'
import { parsePseudoStyles } from './utils'

const systemProps = compose(
  layout,
  color,
  space,
  background,
  border,
  borderRadius,
  grid,
  position,
  shadow,
  typography,
  flexbox,
  propsConfig
)

const CPseudoBox = {
  name: 'CPseudoBox',
  inject: ['$chakraTheme'],
  props: {
    as: {
      type: [String, Object],
      default: () => 'div'
    },
    to: [String, Object],
    ...styleProps
  },
  computed: {
    theme () {
      return this.$chakraTheme()
    }
  },
  render (h) {
    const { as, to, ...cleanedStyleProps } = this.$props
    const { pseudoStyles, baseProps } = parsePseudoStyles(cleanedStyleProps)
    const baseStyles = systemProps({ ...baseProps, theme: this.theme })
    const _pseudoStyles = __css(pseudoStyles)(this.theme)

    return h(CBox, {
      class: css({ ...baseStyles, ..._pseudoStyles }),
      props: {
        as,
        to
      }
    }, this.$slots.default)
  }
}

export default CPseudoBox
