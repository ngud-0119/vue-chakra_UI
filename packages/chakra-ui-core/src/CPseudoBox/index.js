import CPseudoBox from './CPseudoBox'
export default CPseudoBox
export * from './CPseudoBox'
