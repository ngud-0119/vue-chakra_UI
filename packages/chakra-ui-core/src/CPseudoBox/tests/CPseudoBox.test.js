import CPseudoBox from '../'
import { render, getTagName, screen } from '@/tests/test-utils'

const renderComponent = (props) => {
  const inlineAttrs = (props && props.inlineAttrs) || ''
  const base = {
    components: { CPseudoBox },
    template: `<CPseudoBox data-testid="pseudobox" ${inlineAttrs}>Box Works</CPseudoBox>`,
    ...props
  }
  return render(base)
}

it('should render correctly', () => {
  const { asFragment } = renderComponent()
  expect(asFragment()).toMatchSnapshot()
})

it('should change the style', () => {
  const inlineAttrs = `
  d="flex" :w="['auto']" px="5" py="5" shadow="lg"
  my="5" mb="5" rounded="sm" font-family="body"
  background-color="blue.200" color="blue.700"`
  const { asFragment } = renderComponent({ inlineAttrs })

  expect(asFragment()).toMatchSnapshot()
})

it.each`
  as
  ${'header'}
  ${'p'}}
`(
  'should display Box with type as $as',
  ({ as }) => {
    const inlineAttrs = `as=${as}`
    const { asFragment } = renderComponent({ inlineAttrs })
    expect(getTagName(screen.getByTestId('pseudobox'))).toEqual(as)
    expect(asFragment()).toMatchSnapshot()
  }
)

// TODO: How to test _hover or focus styles?
