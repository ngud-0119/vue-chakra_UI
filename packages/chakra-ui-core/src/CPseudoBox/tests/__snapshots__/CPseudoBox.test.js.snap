// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should change the style 1`] = `
<DocumentFragment>
  .emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  width: auto;
  -webkit-padding-start: var(--space-5);
  padding-inline-start: var(--space-5);
  -webkit-padding-end: var(--space-5);
  padding-inline-end: var(--space-5);
  padding-top: var(--space-5);
  padding-bottom: var(--space-5);
  box-shadow: var(--shadows-lg);
  margin-top: var(--space-5);
  margin-bottom: var(--space-5);
  border-radius: var(--radii-sm);
  font-family: var(--fonts-body);
  background-color: var(--colors-blue-200);
  color: var(--colors-blue-700);
}

<div
    class="emotion-0"
    data-chakra-component="CPseudoBox"
    data-testid="pseudobox"
  >
    Box Works
  </div>
</DocumentFragment>
`;

exports[`should display Box with type as header 1`] = `
<DocumentFragment>
  <header
    class="emotion-0"
    data-chakra-component="CPseudoBox"
    data-testid="pseudobox"
  >
    Box Works
  </header>
</DocumentFragment>
`;

exports[`should display Box with type as p 1`] = `
<DocumentFragment>
  <p
    class="emotion-0"
    data-chakra-component="CPseudoBox"
    data-testid="pseudobox"
  >
    Box Works
  </p>
</DocumentFragment>
`;

exports[`should render correctly 1`] = `
<DocumentFragment>
  <div
    class="emotion-0"
    data-chakra-component="CPseudoBox"
    data-testid="pseudobox"
  >
    Box Works
  </div>
</DocumentFragment>
`;
